subprojects/core/src/main/java/org/gradle/api/internal/attributes/DefaultMutableAttributeContainer.java

```diff
@@ -131,17 +131,22 @@ private void assertAttributeValueIsNotNull(@Nullable Object value) {
 
     @Override
     public <T> T getAttribute(Attribute<T> key) {
-        T attribute = state.getAttribute(key);
-        if (attribute == null && lazyAttributes.containsKey(key)) {
-            attribute = realizeLazyAttribute(key);
+        Isolatable<?> value = attributes.get(key);
+        if (value == null) {
+            if (lazyAttributes.containsKey(key)) {
+                return realizeLazyAttribute(key);
+            } else {
+                return null;
+            }
+        } else {
+            return Cast.uncheckedCast(value.isolate());
         }
-        return attribute;
     }
 
     @Override
     public ImmutableAttributes asImmutable() {
         realizeAllLazyAttributes();
-        return state;
+        return immutableAttributesFactory.fromMap(attributes);
```

Comment:
I am concerned with this change here

We no longer cache the immutable version of a mutable container

Each time we call `asImmutable`, we re-construct the immutable version of the container.

Attribute container construction is relatively expensive

Reply:
I see you tried running perf tests but none of those seemed to execute far enough to run any tests